PlayFabSDK/source/Qos/PlayFabQosApi.cs

```
@@ -80,30 +98,56 @@ private async Task InitializeQoSServerList()
 
-            await Task.WhenAll(asyncPingResults);
-            var results = new List<QosRegionResult>(asyncPingResults.Count);
-            foreach (Task<QosRegionResult> asyncPingResult in asyncPingResults)
-            {
-                results.Add(asyncPingResult.Result);
-            }
+            ConcurrentBag<int> seeds = new ConcurrentBag<int>(Enumerable.Range(0, pingsPerRegion)
+                .Select(i => i * dataCenterMap.Count / pingsPerRegion));
+
+            Task[] pingWorkers = Enumerable.Range(0, degreeOfParallelism).Select(
+                i => PingWorker(regionPingers, seeds)).ToArray();
 
+            await Task.WhenAll(pingWorkers);
+
+            List<QosRegionResult> results = regionPingers.Select(x => x.GetResult()).ToList();
             results.Sort((x, y) => x.LatencyMs.CompareTo(y.LatencyMs));
 
-            return results;
+            QosErrorCode resultCode = QosErrorCode.Success;
+            string errorMessage = null;
+            if (results.All(x => x.ErrorCode == (int) QosErrorCode.NoResult))
+            {
+                resultCode = QosErrorCode.NoResult;
+                errorMessage = "No valid results from any QoS server";
+            }
+            
+            return new QosResult()
+            {
+                ErrorCode = (int) resultCode,
+                RegionResults = results,
+                ErrorMessage = errorMessage
+            };
+        }
+        
+        
+        private async Task PingWorker(RegionPinger[] regionPingers, IProducerConsumerCollection<int> seeds)
+        {
+            while(seeds.TryTake(out int seed))
+            {
+                for (int i = 0; i < regionPingers.Length; i++)
+                {
+                    int index = (i + seed) % regionPingers.Length;
+                    await regionPingers[index].PingAsync();
+                }
+            }
         }
 
         private async Task SendResultsToPlayFab(QosResult result)
```

It might be a good idea to rename this to "SendSuccessfulQosResultsToPlayFab"
Also... We don't give the developer/customer a way to opt out of this callback, and that's not in line with our Data 101 policies. #Resolved

---

It's also odd that we don't seem to want to report a failure to get Qos data back as a PlayStream event...

Finally, I totally accept that none of this feedback is part of your PR, but something I notice while I'm here. #ByDesign

---

Failures are tracked in other ways. This is specifically to collect data on ping times for the developer to be able to visualize how their players are doing.

---
In reply to: [376130054](https://github.com/PlayFab/CSharpSDK/pull/111#discussion_r376130054) [](ancestors = 376130054)